### PR TITLE
Fixed a number of issues related to parsing {(P|A)ts?()} in a multi-line template

### PR DESCRIPTION
**Pull request by RackerWilliams**

### CODE REVIEW

File: core/src/main/resources/xq/mapping2XML.xq [hunk 1]
```diff
@@ -32,8 +33,8 @@ declare function mapping:convertNonMatchValue($in as xs:string) as xs:string {
 
 declare function mapping:convertValue($in as xs:string) as xs:string {
   let $strings as xs:string* :=
-    for $pm in analyze-string($in,'\{.*\}')//(fn:match|fn:non-match) return if (local-name($pm) = 'match') then mapping:convertMatchValue(string($pm))
-                                                                                                           else mapping:convertNonMatchValue(string($pm))
+    for $pm in analyze-string($in,'\{.*?\}','ms')//(fn:match|fn:non-match) return if (local-name($pm) = 'match') then mapping:convertMatchValue(string($pm))
+                                                                                                                            else mapping:convertNonMatchValue(string($pm))
```
Comment (dmnjohns) [inline core/src/main/resources/xq/mapping2XML.xq]: Weird spacing change on this line.

File: core/src/main/resources/xq/mapping2JSON.xq [hunk 2]
```diff
@@ -29,7 +29,7 @@ declare function mapping:convertAttributeValues($tokenized as xs:string*, $match
 };
 
 declare function mapping:tokenizeListValues($in as xs:string, $name as xs:string, $flags as xs:string*) as xs:string* {
-  let $asResult := analyze-string($in,'\{(A|P)ts?\(.*?\)\}')//(fn:match|fn:non-match)
+  let $asResult := analyze-string($in,'\{(A|P)ts?\(.*?\)\}','ms')//(fn:match|fn:non-match)
```
Comment (dmnjohns) [inline core/src/main/resources/xq/mapping2JSON.xq]: Does multi-line mode (i.e., the `m` flag) actually change anything if you aren't using the `^` or `$` meta-characters? The docs (https://www.w3.org/TR/xpath-functions-31/#flags) lead me to believe it doesn't, but your code leads me to believe it does.

File: core/src/main/resources/xq/mapping2XML.xq [hunk 3]
```diff
@@ -20,10 +20,11 @@ declare variable $defaultAttributes as xs:string* := ('name','email','expire','d
 declare variable $remoteMultiValues as xs:string* := ('whitelist','blacklist','notAnyOf','anyOneOf');
 
 declare function mapping:convertMatchValue($in as xs:string) as xs:string {
+
  (:
    Only the following should not be escaped :  {At(...)}, {Ats(...)}, {Pt(...)}, {Pts(...)}.
  :)
- if (matches($in,'^\{(A|P)ts?\(.*\)\}$')) then $in else mapping:convertNonMatchValue($in)
+ if (matches($in,'\{(A|P)ts?\(.*?\)\}','ms')) then $in else mapping:convertNonMatchValue($in)
```
Comment (dmnjohns) [inline core/src/main/resources/xq/mapping2XML.xq]: Hm, so dropping the `^` and `$` characters means this becomes a loose regex match, so if any substring matches then the entire string matches. That would mean that the following matches:
```
{At(foo)}
{Pt(bar)}
{Pts()}
```

Is that desired?

File: core/src/main/resources/xsl/mapping.xsl [hunk 4]
```diff
@@ -436,54 +416,83 @@
...
-                <xsl:choose>
-                    <xsl:when test="(local-name($parent/..) = 'user') and
-                                    (namespace-uri($parent/..)='http://docs.rackspace.com/identity/api/ext/MappingRules') ">
-                        <xsl:sequence select="mapping:defaultForName(local-name($parent))"/>
-                    </xsl:when>
-                    <xsl:otherwise>
-                        <xsl:message terminate="yes">[ERROR] The attribute <xsl:value-of select="concat(name($parent/..),'/',name($parent))"/> does not have a default value.</xsl:message>
-                    </xsl:otherwise>
-                </xsl:choose>
+            <xsl:when test="matches($part,'\{D\}','ms')">
```
Comment (dmnjohns) [inline core/src/main/resources/xsl/mapping.xsl]: The matching here also got looser. We went from part must equal `{D}` to part must contain `{D}`. The following part would now match:
```
some{D}
```

Is that desired?

Comment (dmnjohns) [reply core/src/main/resources/xsl/mapping.xsl]: Also, neither the `m` nor `s` flags seem to be used in this case.

File: core/src/main/resources/xsl/mapping.xsl [hunk 5]
```diff
@@ -436,54 +416,83 @@
...
+                        </xsl:choose>
+                    </xsl:matching-substring>
+                    <xsl:non-matching-substring>
+                        <xsl:sequence select="mapping:mapAttribute($parent,.,$remoteMappers)"/>
+                    </xsl:non-matching-substring>
+                </xsl:analyze-string>
             </xsl:when>
-            <xsl:when test="matches($part,'\{D\(.*\)\}')">
-                <xsl:analyze-string select="$part" regex="\{{D\((.*)\)\}}">
+            <xsl:when test="matches($part,'\{D\(.*?\)\}')">
```
Comment (dmnjohns) [inline core/src/main/resources/xsl/mapping.xsl]: I wonder why this `matches` is not passed the `ms` flags, considering that the `s` flag would apply here. So the following would not match:
```
{D(
foobar
)}
```
even though the following would match below:
```
{At(
foobar
)}
```

Is that desired?

File: core/src/main/resources/xsl/mapping.xsl [hunk 6]
```diff
@@ -436,54 +416,83 @@
...
+                <xsl:analyze-string select="$part" regex="\{{D\((.*?)\)\}}">
```
Comment (dmnjohns) [inline core/src/main/resources/xsl/mapping.xsl]: Ooh, attribute value templates playing a role here and making escaping just a bundle of fun.

File: core/src/main/resources/xsl/mapping.xsl [hunk 7]
```diff
@@ -515,7 +524,7 @@
         <xsl:param name="name" as="xs:string"/>
         <xsl:variable name="attribDefault" as="element()?" select="$defaults/mapping:attribute-defaults/mapping:attribute[@name=$name]"/>
         <xsl:if test="empty($attribDefault)">
-            <xsl:message terminate="yes">[ERROR] No default value for attribute <xsl:value-of select="$name"/></xsl:message>
+            <xsl:message terminate="yes">[ERROR] No default value for attribute   '<xsl:value-of select="$name"/>'</xsl:message>
```
Comment (dmnjohns) [inline core/src/main/resources/xsl/mapping.xsl]: I see the single quotes that were added, but why the leading spaces?

File: core/src/main/resources/xq/mapping2XML.xq (hunk 3 above)
Comment (dmnjohns) [reply core/src/main/resources/xq/mapping2XML.xq]: Considering the test which includes `{At(FirstName)} {At(LastName)} <{D(name)}@{D(domain)}.rackspace.com>`, I'm guessing that yes, that is intended.

File: core/src/main/resources/xsl/mapping.xsl (hunk 4 above)
Comment (dmnjohns) [reply core/src/main/resources/xsl/mapping.xsl]: Considering the same test as above, the looser matching probably was intentional to support multiple references at once.

File: core/src/main/resources/xq/mapping2JSON.xq (hunk 2 above)
Comment (RackerWilliams) [reply core/src/main/resources/xq/mapping2JSON.xq]: I suppose m doesn't have an effect, but leaving it there to remind me that we're multiline if I make changes.  We are now looking at things in the middle of the string so $ and ^ change meaning.  s certainly is relevant here as it changes the meaning of the ..

File: core/src/main/resources/xq/mapping2XML.xq (hunk 3 above)
Comment (RackerWilliams) [reply core/src/main/resources/xq/mapping2XML.xq]: That is correct and the right behavior.

File: core/src/main/resources/xsl/mapping.xsl (hunk 4 above)
Comment (RackerWilliams) [reply core/src/main/resources/xsl/mapping.xsl]: m and s are there to remind me that I'm in multi-line and things are in the middle of a string. 
same{D} should match -- we are templating.

File: core/src/main/resources/xq/mapping2JSON.xq (hunk 2 above)
Comment (dmnjohns) [reply core/src/main/resources/xq/mapping2JSON.xq]: Fair enough.

File: core/src/main/resources/xsl/mapping.xsl (hunk 5 above)
Comment (RackerWilliams) [reply core/src/main/resources/xsl/mapping.xsl]: Correct.  Whitespace matters in attribute names -- so you need to be allowed to enter it there -- I doubt it'll make sens to do the At( example.  The attribute name in {D()} is always a single standard field name.  So decided to match it that way.

File: core/src/main/resources/xsl/mapping.xsl (hunk 6 above)
Comment (RackerWilliams) [reply core/src/main/resources/xsl/mapping.xsl]: You got that right :-)

File: core/src/main/resources/xsl/mapping.xsl (hunk 7 above)
Comment (RackerWilliams) [reply core/src/main/resources/xsl/mapping.xsl]: Helped me debug.  I was incorrectly catching the name and it had spaces in there, couldn't see from the error message so added the quotes.  Decided to leave there for future.

**Pull request merged**